### ISSUE TITLE
docker_network is not applied correctly on first run

### ISSUE THREAD

**Issue by jhaezebr** [puppetlabs/puppetlabs-docker#703]
## Describe the Bug
When I try to make a docker_network on a clean install (without docker already present on the system), puppet apply gives the following error:
`Error: Could not find a suitable provider for docker_network`
If the apply command is repeated a second time, the error does not occur.

## Expected Behavior
The docker network is created.

## Steps to Reproduce
On a clean install with puppet 6:
`puppet module install puppetlabs/docker`
`puppet apply -e 'include docker; docker_network {'test_net': ensure => present,}'`

## Environment
 - Version v3.13.0
 - Platform Ubuntu20.04
 - Puppet 6.19.1

## Additional Context
I don't have much experience with puppet or ruby, but here is what I think is happening:

From what I've seen the problem stems from the following code [here](https://github.com/puppetlabs/puppetlabs-docker/blob/2320dec1c94b6c7dea0ecb7f44db264153c1c014/lib/puppet/provider/docker_network/ruby.rb#L11).
Because docker is not present on the system when the provider code is run, the `command(:dockercmd)` function returns nil. The :docker command is therefore created without the proper binary path or name. When the provider needs to create the docker_network, it fails.
I have tried a few things and removing :dockercmd and replacing the has_command function with `has_command(:docker, 'docker')` seems to work.

Because the code is the same, I guess that this bug will be present in all providers.

**Comment by shoddyguard** [comment]
+1 also seeing this in my environment.

* Version v3.13.1
* Platform Ubuntu20.04
* Puppet 6.21

**Comment by mendhak** [comment]
@shoddyguard  @jhaezebr  
Did you ever find a way to work around this problem?  I'm on Puppet 7.6.1 (Ubuntu 20.04) and I'm seeing this problem with docker network and docker compose calls.  It only works on the 2nd run.

**Comment by jhaezebr** [comment]
@mendhak  Not really, I'm just running puppet twice on a new install

**Comment by mendhak** [comment]
OK thanks, I'm having to run it twice too.  

Sadly the code changes didn't work for me either.  I tried following your post, and replaced these lines according to my understanding:

```
  commands dockercmd: 'docker'

  has_command(:docker, command(:dockercmd)) do
    environment(HOME: '/root')
  end
```

with 

```
  has_command(:docker, 'docker') do
    environment(HOME: '/root')
  end
```

but it still required running twice.   I hope someone from puppetlabs will notice this bug

**Comment by mendhak** [comment]
>  If this issue is still important to you please drop a comment below and we will add this to our backlog to complete

Well yes this is still important.  It would be good to have a non-failing first run.

**Comment by canihavethisone** [comment]
I use the sed script below to fix this issue, noting that it uses fixed Linux paths (not windows).  I attempted to use an exec within the has_command thus, but it returns nil:
` has_command(:dockercompose, `which docker-compose`) do`

```
#!/bin/bash

# The following sed actions will patch an issue with the puppetlabs-docker module that requires 2 puppet runs

# Change path to your modules dir
target_path='/etc/puppetlabs/code/environments/production/modules'

## --- DO NOT EDIT BELOW THIS LINE --- ##

target_file_1='docker/lib/puppet/provider/docker_compose/ruby.rb'
target_file_2='docker/lib/puppet/provider/docker_network/ruby.rb'
target_file_3='docker/lib/puppet/provider/docker_stack/ruby.rb'
target_file_4='docker/lib/puppet/provider/docker_volume/ruby.rb'

find_1="has_command(:docker, command(:dockercmd)) do"
replace_1="has_command(:docker, '/usr/bin/docker') do"

find_2="has_command(:docker_compose, command(:dockercompose)) do"
replace_2="has_command(:dockercompose, '/usr/local/bin/docker-compose') do"


sed -i "s;$find_1;$replace_1;g" \
    $target_path/{$target_file_1,$target_file_2,$target_file_3,$target_file_4}

sed -i "s;$find_2;$replace_2;g" \
    $target_path/$target_file_1
```

I still find that a second puppet run is required when `docker::selinux_enabled: true` as it performs `/etc/systemd/system/docker.service.d/service-overrides.conf]/seltype: seltype changed 'systemd_unit_file_t' to 'container_unit_file_t'` and also on any docker service files in `/etc/systemd/system/`

**Comment by canihavethisone** [comment]
I have created a pull request at https://github.com/puppetlabs/puppetlabs-docker/pull/833.  Can other users please test and verify these changes?  Thanks.

**Comment by chelnak** [comment]
Fixed by #833
